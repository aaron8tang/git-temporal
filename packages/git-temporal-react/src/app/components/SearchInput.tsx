import React from 'react';
import { style } from 'app/styles';

import { debounce } from 'app/utilities/debounce';
import { SearchIcon } from 'app/components/SearchIcon';

export interface SearchInputProps {
  value: string;
  onChange: (value) => void;
  onClear: () => void;
  onFocus?: (evt) => void;
  onBlur?: (evt) => void;
  onKeyDown?: (evt) => void;
  placeholder?: string;
  style?: object;
}

interface SearchInputState {
  value: string;
}

const containerStyle = {
  _extends: ['flexRows', 'borderedPanel', 'normalText', 'selectable'],
  position: 'relative',
  padding: 5,
};
const searchInputStyle = {
  flexGrow: 1,
  border: 'none',
  fontSize: 14,
  margin: '0px 5px',
};
const clearIconStyle = {
  _extends: ['smallerText'],
  paddingTop: 3,
  flexGrow: 0,
  cursor: 'pointer',
};
const searchIconStyle = {
  marginTop: 2,
};

export class SearchInput extends React.Component<
  SearchInputProps,
  SearchInputState
> {
  static displayName = 'SearchInput';
  readonly state: SearchInputState = { value: '' };

  private inputRef;
  private debouncedCallChangeCallback;

  constructor(props) {
    super(props);
    this.debouncedCallChangeCallback = debounce(this.callChangeCallback, 750);
  }

  componentDidMount() {
    this.inputRef && this.inputRef.focus();
    if (this.props.value !== this.state.value) {
      this.setState({ value: this.props.value });
    }
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.value !== this.props.value &&
      this.props.value !== this.state.value
    ) {
      this.setState({ value: this.props.value });
    }
  }

  render() {
    const { onClear, onFocus, onBlur, onKeyDown, placeholder } = this.props;
    return (
      <div style={style(containerStyle, this.props.style)}>
        <SearchIcon height={16} width={16} style={style(searchIconStyle)} />
        <input
          type="text"
          style={searchInputStyle}
          value={this.state.value}
          placeholder={placeholder}
          onChange={this.onInputChange}
          onFocus={onFocus}
          onBlur={onBlur}
          onKeyDown={onKeyDown}
          ref={input => {
            this.inputRef = input;
          }}
        />
        <div style={style(clearIconStyle)} onClick={onClear}>
          X
        </div>
      </div>
    );
  }

  onInputChange = evt => {
    const newValue = evt.target.value;
    this.setState({ value: newValue });
    this.debouncedCallChangeCallback(newValue);
  };

  callChangeCallback = value => {
    this.props.onChange(value);
  };
}
