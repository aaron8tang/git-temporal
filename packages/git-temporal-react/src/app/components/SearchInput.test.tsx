import React from 'react';
import { shallow } from 'enzyme';

import { SearchInput } from './SearchInput';

describe('components/SearchInput', () => {
  describe('when rendered with value', () => {
    let wrapper;
    const onChangeMock = jest.fn();
    const onClearMock = jest.fn();
    beforeAll(() => {
      wrapper = shallow(
        <SearchInput
          value={'test value'}
          onChange={onChangeMock}
          onClear={onClearMock}
        />
      );
    });

    test('it should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('and then changing input, should call onChange', done => {
      const event = {
        preventDefault() {},
        target: { value: 'the-value' },
      };
      wrapper.find('input').simulate('change', event);
      // the component should debounce the input and change callback
      expect(onChangeMock).toHaveBeenCalledTimes(0);
      setTimeout(() => {
        expect(onChangeMock).toBeCalledWith('the-value');
        done();
      }, 1000);
    });
  });
});
