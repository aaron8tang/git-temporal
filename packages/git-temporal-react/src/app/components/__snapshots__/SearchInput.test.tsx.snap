// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SearchInput when rendered with value it should match snapshot 1`] = `
<div
  style={
    Object {
      "background": "white",
      "border": "1px solid lightskyblue",
      "borderRadius": 3,
      "color": "#333333",
      "cursor": "pointer",
      "display": "flex",
      "flexDirection": "row",
      "fontSize": 13,
      "fontWeight": "normal",
      "marginRight": 10,
      "padding": 5,
      "position": "relative",
      "userSelect": "none",
    }
  }
>
  <SearchIcon
    height={16}
    style={
      Object {
        "marginTop": 2,
      }
    }
    width={16}
  />
  <input
    onChange={[Function]}
    style={
      Object {
        "border": "none",
        "flexGrow": 1,
        "fontSize": 14,
        "margin": "0px 5px",
      }
    }
    type="text"
    value="test value"
  />
  <div
    onClick={[MockFunction]}
    style={
      Object {
        "color": "#333333",
        "cursor": "pointer",
        "flexGrow": 0,
        "fontSize": 11,
        "fontWeight": "normal",
        "paddingTop": 3,
      }
    }
  >
    X
  </div>
</div>
`;
